src/middleware/csrf.ts

```diff
@@ -0,0 +1,20 @@
+import { Response, Request, NextFunction } from "express";
+import { generateCSRFToken, verifyCSRFToken } from '../services/tokenService';
+import { Cache } from '../datasources/cache';
+
+export async function csrfMiddleware(req: Request, res: Response, next: NextFunction) {
+  const cache = Cache.getInstance();
+
+  // Only worry about the CSRF token if the caller is performing a POST, PUT, PATCH, DELETE
+  if (req.method !== 'GET' && req.method !== 'HEAD' && req.method !== 'OPTIONS') {
+    const token = req.headers['x-csrf-token'] as string;
+    if (!token || !(await verifyCSRFToken(cache, token))) {
```

Comment:
L11 - If the frontend initially calls the "/apollo-csrf" endpoint to get a CSRF token, will the user get a 'Invalid CSRF token' error instead of getting a new CSRF token?

Reply:
No. The "/apollo-csrf" endpoint is a GET, so the CSRF token verification doesn't get triggered